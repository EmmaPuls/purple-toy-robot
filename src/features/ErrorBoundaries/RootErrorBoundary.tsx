import { css } from "@emotion/css";
import defaultTo from "lodash/defaultTo";
import { Component, ErrorInfo, ReactNode } from "react";
import ErrorPage from "./ErrorPage";

interface Props {
  children: ReactNode;
  themeLoaded?: boolean;
}

interface State {
  hasError: boolean;
  themeLoaded: boolean;
  title: string;
  body: string;
}

/**
 * Component representing an error boundary.
 * Can render if theme is not available.
 * Will render with styles if theme is available.
 */
class RootErrorBoundary extends Component<Props, State> {
  state: State = {
    hasError: false,
    themeLoaded: false,
    title: "Oh no!\nSomething went wrong",
    body: "Reload the page to continue",
  };

  constructor(props: Props) {
    super(props);
    this.state.themeLoaded = defaultTo(props.themeLoaded, false);
  }

  static getDerivedStateFromError(error: any) {
    return { hasError: true };
  }

  componentDidCatch(error: any, errorInfo: ErrorInfo) {
    console.error(error, errorInfo);
  }

  render() {
    if (this.state.hasError) {
      if (this.state.themeLoaded) {
        return (
          <ErrorPage
            title={this.state.title}
            body={this.state.body}
            themed={this.state.themeLoaded}
          />
        );
      } else {
        return (
          <>
            <h1 className={css(`padding-left:20px; white-space:pre-wrap`)}>{this.state.title}</h1>
            <p className={css(`padding-left:20px`)}>{this.state.body}</p>
          </>
        );
      }
    }

    return this.props.children;
  }
}

export default RootErrorBoundary;
