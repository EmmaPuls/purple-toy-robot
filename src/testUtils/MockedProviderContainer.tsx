import { Theme, ThemeProvider } from "@emotion/react";
import { defaultTo } from "lodash";
import React, { FC } from 'react';
import myTheme from "../theme";

type MockedProviderContainerProps = {
  theme?: Theme;
  children: React.ReactNode;
};

const MockedProviderContainer: FC<MockedProviderContainerProps> = ({ children, theme }) => {
  return (
    <ThemeProvider theme={defaultTo(theme, myTheme)}>
      {children}
    </ThemeProvider>
  );
};

export default MockedProviderContainer;
