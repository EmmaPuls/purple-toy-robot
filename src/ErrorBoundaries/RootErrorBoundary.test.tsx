import { render } from "@testing-library/react";
import MockedProviderContainer from "testUtils/MockedProviderContainer";
import RootErrorBoundary from "./RootErrorBoundary";

describe("RootErrorBoundary", () => {
  const ThrowErrorTest = () => {
    throw new Error("Error");
  };

  it("should render error boundary without theme by default", () => {
    render(
      <RootErrorBoundary>
        <ThrowErrorTest />
      </RootErrorBoundary>
    );

    expect(document.body).toMatchSnapshot();
  });

  it("should render error boundary with theme if requested", () => {
    render(
      <MockedProviderContainer>
        <RootErrorBoundary themeLoaded>
          <ThrowErrorTest />
        </RootErrorBoundary>
      </MockedProviderContainer>
    );

    expect(document.body).toMatchSnapshot();
  });

  it("should render error boundary without theme if error in theme", () => {
    render(
      <RootErrorBoundary>
        <RootErrorBoundary themeLoaded>
          <ThrowErrorTest />
        </RootErrorBoundary>
      </RootErrorBoundary>
    );

    expect(document.body).toMatchSnapshot();
  });
});
