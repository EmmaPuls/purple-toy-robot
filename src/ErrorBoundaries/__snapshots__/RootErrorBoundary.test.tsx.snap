// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RootErrorBoundary should render error boundary with theme if requested 1`] = `
<body>
  <div>
    <div
      class="css-1fovf95"
    >
      <div
        class="css-1ay46ss"
      >
        <h1
          class="css-1kxrhf3"
        >
          Oh no!
Something went wrong
        </h1>
        <p
          class="css-1x1uhd2"
        >
          Reload the page to continue
        </p>
      </div>
    </div>
  </div>
</body>
`;

exports[`RootErrorBoundary should render error boundary without theme by default 1`] = `
<body>
  <div>
    <h1
      class="css-1u8eb7y"
    >
      Oh no!
Something went wrong
    </h1>
    <p
      class="css-1tyyxc3"
    >
      Reload the page to continue
    </p>
  </div>
</body>
`;

exports[`RootErrorBoundary should render error boundary without theme if error in theme 1`] = `
<body>
  <div>
    <h1
      class="css-1u8eb7y"
    >
      Oh no!
Something went wrong
    </h1>
    <p
      class="css-1tyyxc3"
    >
      Reload the page to continue
    </p>
  </div>
</body>
`;
